cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java

```diff
@@ -0,0 +1,191 @@
+   */
+  void addMetadata(final Id.NamespacedId entityId, final Map<String, String> metadata) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        for (Map.Entry<String, String> entry : metadata.entrySet()) {
+          input.businessMds.createBusinessMetadata(entityId, entry.getKey(), entry.getValue());
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Adds tags for the specified {@link Id.NamespacedId}.
+   */
+  void addTags(final Id.NamespacedId entityId, final String ... tagsToAdd) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        Map<String, String> businessMetadata = input.businessMds.getBusinessMetadata(entityId);
+        Iterable<String> existingTags = getTags(businessMetadata);
+        Iterable<String> newTags = Iterables.concat(existingTags, Arrays.asList(tagsToAdd));
+        input.businessMds.createBusinessMetadata(entityId, "tags", Joiner.on(",").join(newTags));
+        return null;
+      }
+    });
+  }
+
+  /**
+   * @return the metadata for the specified {@link Id.NamespacedId}
+   */
+  Map<String, String> getMetadata(final Id.NamespacedId entityId) {
+    return txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Map<String, String>>() {
+      @Override
+      public Map<String, String> apply(BusinessMdsIterable input) throws Exception {
+        return input.businessMds.getBusinessMetadata(entityId);
+      }
+    });
+  }
+
+  /**
+   * @return the tags for the specified {@link Id.NamespacedId}
+   */
+  Iterable<String> getTags(final Id.NamespacedId entityId) {
+    return txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Iterable<String>>() {
+      @Override
+      public Iterable<String> apply(BusinessMdsIterable input) throws Exception {
+        Map<String, String> businessMetadata = input.businessMds.getBusinessMetadata(entityId);
+        return getTags(businessMetadata);
+      }
+    });
+  }
+
+  /**
+   * Removes all metadata for the specified {@link Id.NamespacedId}.
+   */
+  void removeMetadata(final Id.NamespacedId entityId) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        input.businessMds.removeMetadata(entityId);
+        return null;
+      }
+    });
+  }
+
+  void removeMetadata(final Id.NamespacedId entityId, final String ... keys) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        input.businessMds.removeMetadata(entityId, keys);
+        return null;
```

Comment:
This is a good example of better not to use `Transactional`. The `TransactionExecutor` support executing a `Procedure` that doesn't need to return anything.